File: app/services/versioning.py
"""Company/scenario creation, thesis amendment, and version diffing
(BUILD_PLAN.md §1.3, §6; per-user scenarios per ADR-026).

thesis_versions is append-only (enforced by a DB trigger - see the P0
migration); this module only ever INSERTs new version rows and repoints
scenario.current_version_id, never UPDATEs an existing version.
"""
from typing import Any

from sqlalchemy import func, select
from sqlalchemy.exc import IntegrityError
from sqlalchemy.orm import Session

from app.config import ANALYST_NAME
from app.models import BroadIndustry, Company, KillTrigger, SpecificNiche, ThesisScenario, ThesisVersion
from app.schemas.thesis import ThesisCreate, ThesisData
from app.services.scenarios import ScenarioNotFoundError, get_scenario_optional


class NotFoundError(Exception):
    pass


class TaxonomyError(Exception):
    pass


class AlreadyExistsError(Exception):
    pass


def _resolve_taxonomy(
    db: Session, broad_industry_name: str, specific_niche_name: str
) -> tuple[BroadIndustry, SpecificNiche]:
    industry = db.scalar(select(BroadIndustry).where(BroadIndustry.name == broad_industry_name))
    if industry is None:
        raise TaxonomyError(f"unknown broad_industry '{broad_industry_name}'")
    niche = db.scalar(
        select(SpecificNiche).where(
            SpecificNiche.broad_industry_id == industry.id,
            SpecificNiche.name == specific_niche_name,
        )
    )
    if niche is None:
        raise TaxonomyError(
            f"unknown specific_niche '{specific_niche_name}' under '{broad_industry_name}'; "
            "propose it via POST /taxonomy/niches first"
        )
    return industry, niche


def _write_kill_triggers(db: Session, version_id: int, thesis_data: ThesisData) -> None:
    for trigger in thesis_data.what_can_kill_it:
        db.add(
            KillTrigger(
                version_id=version_id,
                label=trigger.label,
                metric_key=trigger.metric_key,
                operator=trigger.operator,
                threshold=trigger.threshold,
                severity=trigger.severity,
                action=trigger.action,
                grace_periods=trigger.grace_periods,
                manual_check=trigger.manual_check,
            )
        )


def create_company(db: Session, payload: ThesisCreate, actor: str = ANALYST_NAME) -> ThesisScenario:
    """Creates a brand-new company + the caller's first scenario on it, OR -
    if the company already exists under someone else's thesis - just starts
    the caller's own new scenario on that existing company (same endpoint,
    same payload shape; classification fields are ignored for an existing
    company since identity is shared and already set). Raises
    AlreadyExistsError only if the caller already has a scenario here.
    """
    company = db.get(Company, payload.company_id)
    if company is not None and get_scenario_optional(db, payload.company_id, actor) is not None:
        raise AlreadyExistsError(f"'{actor}' already has a thesis on company '{payload.company_id}'")

    if company is None:
        industry, niche = _resolve_taxonomy(
            db, payload.classification.broad_industry, payload.classification.specific_niche
        )
        company = Company(
            company_id=payload.company_id,
            name=payload.name,
            broad_industry_id=industry.id,
            specific_niche_id=niche.id,
            operating_model=payload.classification.operating_model,
            currency=payload.classification.currency,
        )
        db.add(company)
        db.flush()

    scenario = ThesisScenario(
        company_id=company.company_id,
        owner=actor,
        label="Thesis",
        status=payload.status,
        status_source="manual",
        last_reviewed=payload.last_reviewed,
    )
    db.add(scenario)
    db.flush()

    version = ThesisVersion(
        company_id=company.company_id,
        scenario_id=scenario.id,
        version_no=1,
        thesis_data=payload.thesis_data.model_dump(mode="json"),
        change_note="initial thesis",
        authored_by=actor,
    )
    db.add(version)
    db.flush()

    _write_kill_triggers(db, version.version_id, payload.thesis_data)

    scenario.current_version_id = version.version_id
    db.commit()
    db.refresh(scenario)
    return scenario


def amend_thesis(
    db: Session, company_id: str, thesis_data: ThesisData, change_note: str, actor: str = ANALYST_NAME
) -> ThesisVersion:
    scenario = get_scenario_optional(db, company_id, actor)
    if scenario is None:
        if db.get(Company, company_id) is None:
            raise NotFoundError(f"company '{company_id}' not found")
        raise ScenarioNotFoundError(f"'{actor}' has no thesis on company '{company_id}' yet - start one first")

    version: ThesisVersion | None = None
    max_attempts = 5
    for attempt in range(max_attempts):
        next_version_no = (
            db.scalar(select(func.max(ThesisVersion.version_no)).where(ThesisVersion.scenario_id == scenario.id))
            or 0
        ) + 1

        version = ThesisVersion(
            company_id=company_id,
            scenario_id=scenario.id,
            version_no=next_version_no,
            thesis_data=thesis_data.model_dump(mode="json"),
            change_note=change_note,
            authored_by=actor,
        )
        db.add(version)
        try:
            db.flush()
            break
        except IntegrityError:
            db.rollback()
            scenario = get_scenario_optional(db, company_id, actor)
            if attempt == max_attempts - 1:
                raise

    _write_kill_triggers(db, version.version_id, thesis_data)
    scenario.current_version_id = version.version_id
    db.commit()
    db.refresh(version)
    return version


def _flatten(obj: Any, prefix: str = "") -> dict[str, Any]:
    flat: dict[str, Any] = {}
    if isinstance(obj, dict):
        for key, value in obj.items():
            flat.update(_flatten(value, f"{prefix}.{key}" if prefix else key))
    elif isinstance(obj, list):
        for i, value in enumerate(obj):
            flat.update(_flatten(value, f"{prefix}[{i}]"))
    else:
        flat[prefix] = obj
    return flat


def diff_versions(v1: ThesisVersion, v2: ThesisVersion) -> list[dict[str, Any]]:
    """Flat dotted-path structural diff between two versions' thesis_data."""
    left = _flatten(v1.thesis_data)
    right = _flatten(v2.thesis_data)
    changes = []
    for path in sorted(set(left) | set(right)):
        old, new = left.get(path), right.get(path)
        if old != new:
            changes.append({"path": path, "old": old, "new": new})
    return changes

File: web/src/lib/server/services/versioning.ts
// Ports app/services/versioning.py
import { and, eq, max } from 'drizzle-orm';
import { db } from '../db';
import {
	broadIndustries,
	companies,
	killTriggers,
	operatingModels,
	specificNiches,
	thesisScenarios,
	thesisVersions
} from '../db/schema';
import type { ThesisCreate, ThesisData } from '../schemas/thesis';
import { getScenarioOptional, ScenarioNotFoundError } from './scenarios';

export class NotFoundError extends Error {}
export class TaxonomyError extends Error {}
export class AlreadyExistsError extends Error {}

async function resolveTaxonomy(broadIndustryName: string, specificNicheName: string) {
	const [industry] = await db.select().from(broadIndustries).where(eq(broadIndustries.name, broadIndustryName)).limit(1);
	if (!industry) throw new TaxonomyError(`unknown broad_industry '${broadIndustryName}'`);
	const [niche] = await db
		.select()
		.from(specificNiches)
		.where(and(eq(specificNiches.broadIndustryId, industry.id), eq(specificNiches.name, specificNicheName)))
		.limit(1);
	if (!niche) {
		throw new TaxonomyError(
			`unknown specific_niche '${specificNicheName}' under '${broadIndustryName}'; propose it via POST /taxonomy/niches first`
		);
	}
	return { industry, niche };
}

async function resolveOperatingModel(name: string) {
	const [model] = await db.select().from(operatingModels).where(eq(operatingModels.name, name)).limit(1);
	if (!model) {
		throw new TaxonomyError(`unknown operating_model '${name}'; propose it via POST /taxonomy/operating-models first`);
	}
	return model;
}

async function writeKillTriggers(versionId: number, thesis: ThesisData) {
	if (!thesis.what_can_kill_it.length) return;
	await db.insert(killTriggers).values(
		thesis.what_can_kill_it.map((t) => ({
			versionId,
			label: t.label,
			metricKey: t.metric_key ?? null,
			operator: t.operator ?? null,
			threshold: t.threshold != null ? String(t.threshold) : null,
			severity: t.severity,
			action: t.action,
			gracePeriods: t.grace_periods,
			manualCheck: t.manual_check
		}))
	);
}

// The form no longer collects a raw company_id - it derives one from
// whichever exchange ticker was supplied (NSE preferred, since that's the
// more commonly quoted one), falling back to a passed-through company_id
// for API/JSON-import callers that still provide it directly.
function deriveCompanyId(payload: ThesisCreate): string {
	if (payload.company_id) return payload.company_id;
	const raw = payload.nse_ticker || payload.bse_ticker;
	if (!raw) throw new TaxonomyError('one of nse_ticker, bse_ticker, or company_id is required');
	return raw.replace(/[^A-Z0-9_]/g, '_').slice(0, 50);
}

export async function createCompany(payload: ThesisCreate, actor: string) {
	const companyId = deriveCompanyId(payload);
	const [existing] = await db.select().from(companies).where(eq(companies.companyId, companyId)).limit(1);
	if (existing && (await getScenarioOptional(companyId, actor))) {
		throw new AlreadyExistsError(`'${actor}' already has a thesis on company '${companyId}'`);
	}

	let company = existing;
	if (!company) {
		const { industry, niche } = await resolveTaxonomy(
			payload.classification.broad_industry,
			payload.classification.specific_niche
		);
		const model = await resolveOperatingModel(payload.classification.operating_model);
		[company] = await db
			.insert(companies)
			.values({
				companyId,
				name: payload.name,
				nseTicker: payload.nse_ticker ?? null,
				bseTicker: payload.bse_ticker ?? null,
				broadIndustryId: industry.id,
				specificNicheId: niche.id,
				operatingModel: model.name,
				currency: payload.classification.currency
			})
			.returning();
	}

	const [scenario] = await db
		.insert(thesisScenarios)
		.values({
			companyId: company.companyId,
			owner: actor,
			label: 'Thesis',
			status: payload.status,
			statusSource: 'manual',
			lastReviewed: payload.last_reviewed
		})
		.returning();

	const [version] = await db
		.insert(thesisVersions)
		.values({
			companyId: company.companyId,
			scenarioId: scenario.id,
			versionNo: 1,
			thesisData: payload.thesis_data,
			changeNote: 'initial thesis',
			authoredBy: actor
		})
		.returning();

	await writeKillTriggers(version.versionId, payload.thesis_data);

	const [updated] = await db
		.update(thesisScenarios)
		.set({ currentVersionId: version.versionId })
		.where(eq(thesisScenarios.id, scenario.id))
		.returning();

	return updated;
}

export async function updateCompanyDetails(
	companyId: string,
	patch: { name?: string; broad_industry?: string; specific_niche?: string; operating_model?: string; currency?: string }
) {
	const [company] = await db.select().from(companies).where(eq(companies.companyId, companyId)).limit(1);
	if (!company) throw new NotFoundError(`company '${companyId}' not found`);

	const values: Partial<typeof companies.$inferInsert> = {};
	if (patch.name !== undefined) values.name = patch.name;
	if (patch.operating_model !== undefined) {
		const model = await resolveOperatingModel(patch.operating_model);
		values.operatingModel = model.name;
	}
	if (patch.currency !== undefined) values.currency = patch.currency;
	if (patch.broad_industry !== undefined || patch.specific_niche !== undefined) {
		const { industry, niche } = await resolveTaxonomy(
			patch.broad_industry ?? (await db.select().from(broadIndustries).where(eq(broadIndustries.id, company.broadIndustryId)).limit(1))[0].name,
			patch.specific_niche ?? (await db.select().from(specificNiches).where(eq(specificNiches.id, company.specificNicheId)).limit(1))[0].name
		);
		values.broadIndustryId = industry.id;
		values.specificNicheId = niche.id;
	}

	const [updated] = await db.update(companies).set(values).where(eq(companies.companyId, companyId)).returning();
	return updated;
}

export async function amendThesis(companyId: string, thesisData: ThesisData, changeNote: string, actor: string) {
	const scenario = await getScenarioOptional(companyId, actor);
	if (!scenario) {
		const [company] = await db.select().from(companies).where(eq(companies.companyId, companyId)).limit(1);
		if (!company) throw new NotFoundError(`company '${companyId}' not found`);
		throw new ScenarioNotFoundError(`'${actor}' has no thesis on company '${companyId}' yet - start one first`);
	}

	const [row] = await db
		.select({ maxNo: max(thesisVersions.versionNo) })
		.from(thesisVersions)
		.where(eq(thesisVersions.scenarioId, scenario.id));
	const nextVersionNo = (row?.maxNo ?? 0) + 1;

	const [version] = await db
		.insert(thesisVersions)
		.values({
			companyId,
			scenarioId: scenario.id,
			versionNo: nextVersionNo,
			thesisData,
			changeNote,
			authoredBy: actor
		})
		.returning();

	await writeKillTriggers(version.versionId, thesisData);
	await db.update(thesisScenarios).set({ currentVersionId: version.versionId }).where(eq(thesisScenarios.id, scenario.id));

	return version;
}

function flatten(obj: unknown, prefix = ''): Record<string, unknown> {
	const flat: Record<string, unknown> = {};
	if (obj !== null && typeof obj === 'object' && !Array.isArray(obj)) {
		for (const [key, value] of Object.entries(obj as Record<string, unknown>)) {
			Object.assign(flat, flatten(value, prefix ? `${prefix}.${key}` : key));
		}
	} else if (Array.isArray(obj)) {
		obj.forEach((value, i) => Object.assign(flat, flatten(value, `${prefix}[${i}]`)));
	} else {
		flat[prefix] = obj;
	}
	return flat;
}

export function diffVersions(v1: { thesisData: unknown }, v2: { thesisData: unknown }) {
	const left = flatten(v1.thesisData);
	const right = flatten(v2.thesisData);
	const paths = [...new Set([...Object.keys(left), ...Object.keys(right)])].sort();
	const changes: { path: string; old: unknown; new: unknown }[] = [];
	for (const path of paths) {
		const oldVal = left[path];
		const newVal = right[path];
		if (JSON.stringify(oldVal) !== JSON.stringify(newVal)) {
			changes.push({ path, old: oldVal ?? null, new: newVal ?? null });
		}
	}
	return changes;
}
